WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,350 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Client.CredentialDependencies
+{
+	public class DependencyGraph
+	{
+		public const int K = ProtocolConstants.CredentialNumber;
+
+		private DependencyGraph(IEnumerable<IEnumerable<long>> initialValues)
+		{
+			var initialValuesImmutable = initialValues.Select(x => x.ToImmutableArray()).ToImmutableArray();
+			if (initialValuesImmutable.Any(x => x.Length != (int)CredentialType.NumTypes))
+			{
+				throw new ArgumentException($"Number of credential values must be {CredentialType.NumTypes}");
+			}
+			if (initialValuesImmutable.Any(x => x.Where(y => y != 0).Select(y => y.CompareTo(0)).Distinct().Count() != 1))
+			{
+				throw new ArgumentException($"Credential values of a node must all have the same sign.");
+			}
+
+			for (CredentialType i = 0; i < CredentialType.NumTypes; i++)
+			{
+				if (initialValuesImmutable.Sum(x => x[(int)i]) < 0)
+				{
+					throw new ArgumentException("Overall balance must not be negative");
+				}
+			}
+
+			// per node entries created in AddNode, querying nodes not in the
+			// graph should result in key errors.
+
+			Successors = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, HashSet<CredentialDependency>>()).ToImmutableArray();
+			Predecessors = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, HashSet<CredentialDependency>>()).ToImmutableArray();
+			EdgeBalances = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, long>()).ToImmutableArray();
```

Comment:
There is a problem here because having `Successors`, `Predecessors` and `EdgeBalances` for each type of credentials means that there is a missing abstraction. I have that here: https://github.com/lontivero/Wasabito/blob/dependency-graph/WalletWasabi/WabiSabi/Client/CredentialDependencies/Graph.cs#L9-L26 and it is clear that it makes the code cleaner and easier to understand.

I am not suggesting to use my abstraction at all but it is more or less clear that this is not clean as it is.

Reply:
I tried to do this but still have a bug: https://github.com/nothingmuch/WalletWasabi/commit/991bd3a868e7a5c93d9150c53a72c93112466639

(see slack DMs from yesterday)

Reply:
see 454ab887663e9514d45dc8405aa4049bc5124289